utils.py

```diff
@@ -889,12 +889,30 @@ def check_auth(user, hutch):
     return False
 
 def check_special(ioc, hutch):
-    lines = open(SPECIAL_FILE % hutch).readlines()
-    lines = [l.strip() for l in lines]
-    for l in lines:
-        if l == ioc:
-            return True
-    return False
+    with open(SPECIAL_FILE % hutch) as fp:
+        lines = fp.readlines()
+        lines = [l.strip() for l in lines]
+        for l in lines:
+            if l == ioc:
+                return True
+        return False
+
+# checks if ioc is marked as toggleable between defined versions of IOCs
+# schema will be ioc_name:permittedversion1,permittedversion2,etc
+def check_special_toggle(ioc, hutch, version):
+    # check that this ioc for this hutch is permissioned specially
+    if not check_special(ioc, hutch):
+        return False
+
+    # check if the permissions extend to toggling firmware versions between the enumerated ones
+    with open(SPECIAL_FILE % hutch) as fp:
+        lines = fp.readlines()
+        for l in lines:
+            permitted_versions = l.split(":")[-1].split(",")
```

Comment:
This is the point at which you need to check l.split(":")[0] to see if it is ioc or not.

Reply:
the above check should be sufficient given the colon does not trip it up

Reply:
I don't believe so.  Suppose the iocmanager.special file was:

ioc1
ioc2: ioc/common/xxx/R1.0, ioc/common/xxx/R2.0

Isn't the command "imgr ioc1 --upgrade ioc/common/xxx/R2.0" going to succeed here?  Or ioc3, for that matter!

Reply:
there is a vulnerability here in that if someone attempts to version the ioc to version name == ioc name it will pass by this logic however, I imagine the deployment will fail later down the line. let me know if this is a permissible assumption

Reply:
I disagree with this as well.  Unless "ioc" appears in one of the comma-separated lists, your current code will not allow an upgrade to that version.

Your problem now is that you aren't ever looking *before* the colon.  If your logic was correct without looking at this, why put in the colon and field before it at all?

Reply:
alright you are totally right; pushed up change to move this all into one function to ensure we are checking that the requested ioc name correlates to a version provided for that ioc name